### PR TITLE
Add a background thread to detect the GC adapter

### PR DESCRIPTION
**Pull request by mathieui**
This should take care of connecting/disconnecting the GC adapter while dolphin is running; previously, dolphin would simply disable the functionality if the adapter wasn’t available on startup.

The thread polls libusb every 500ms if no adapter is found, and it receives a callback from the controller window where it displays the current adapter status (Found/Not Found/Driver not detected).

Why it's not perfect:
- [ ] Due to how it works, we can only detect that an adapter has been disconnected when in-game, because otherwise we don't read from the device. (adding a thread which sole goal would be to make sure an adapter is connected seems overkill)
- [x] An option to enable/disable the thread is possible; but is it necessary? (or maybe I should disable it when the direct connect option is unchecked)

### CODE REVIEW

File: Source/Core/DolphinWX/ControllerConfigDiag.cpp [hunk 1]
```diff
@@ -148,38 +151,59 @@ wxStaticBoxSizer* ControllerConfigDiag::CreateGamecubeSizer()
 	gamecube_static_sizer->AddSpacer(5);
 
 	wxStaticBoxSizer* const gamecube_adapter_group = new wxStaticBoxSizer(wxHORIZONTAL, this, _("GameCube Adapter"));
-	wxBoxSizer* const gamecube_adapter_sizer = new wxBoxSizer(wxHORIZONTAL);
+	wxBoxSizer* const gamecube_adapter_sizer = new wxBoxSizer(wxVERTICAL);
 
 	wxCheckBox* const gamecube_adapter = new wxCheckBox(this, wxID_ANY, _("Direct Connect"));
 	gamecube_adapter->Bind(wxEVT_CHECKBOX, &ControllerConfigDiag::OnGameCubeAdapter, this);
+	m_adapter_status = new wxStaticText(this, wxID_ANY, _(""));
```
Comment (lioncash) [inline Source/Core/DolphinWX/ControllerConfigDiag.cpp]: wxEmptyString

**Comment by skidau** [comment]
Probably should disable the thread when the direct connect option is unchecked.

**Comment by comex** [comment]
There should not be an option to run the thread (as opposed to disabling direct connection entirely), as there would be no point changing it.

Can you please use the [hotplug API](http://libusb.sourceforge.net/api-1.0/hotplug.html) where available?  I don't think it supports Windows, but it supports the other major operating systems.  This avoids the delay of polling.

I haven't really looked at the code, but I don't understand why you would need an additional thread to ensure the device stays connected when you already have one that detects if it is connected.

**Comment by mathieui** [comment]
@comex the Read() thread does not detect anything, it merely performs a dumb read & copy, adding stuff into that loop may add undesirable latency to inputs. However using the hotplug API is a good idea (but still needs a thread on its own, because hotplug needs blocking or frequent calls to handle_events()), so I'm working on integrating both solutions (the first one I coded and the hotplug) and use whichever is available on the platform using the libusb capabilities.

**Comment by JMC47** [comment]
Is there anything else blocking on this?  It's pretty great.

File: Source/Core/Core/HW/SI_GCAdapter.cpp [hunk 2]
```diff
@@ -33,8 +36,15 @@ static int s_controller_payload_size = 0;
 static std::thread s_adapter_thread;
 static Common::Flag s_adapter_thread_running;
 
+static std::thread s_adapter_detect_thread;
+static Common::Flag s_adapter_detect_thread_running;
+
+static std::function<void(void)> s_detect_callback;
```
Comment (lioncash) [inline Source/Core/Core/HW/SI_GCAdapter.cpp]: An `#include <functional>` should also be added.

File: Source/Core/DolphinWX/ControllerConfigDiag.h [hunk 3]
```diff
@@ -64,6 +66,12 @@ class ControllerConfigDiag : public wxDialog
 		SConfig::GetInstance().m_GameCubeAdapter = event.IsChecked();
 		event.Skip();
 	}
+	void OnAdapterRumble(wxCommandEvent& event)
+	{
+		SConfig::GetInstance().m_AdapterRumble = event.IsChecked();
+		event.Skip();
```
Comment (lioncash) [inline Source/Core/DolphinWX/ControllerConfigDiag.h]: Calling skip here isn't necessary.

**Comment by solidshredder** [comment]
There definitely needs to be an option to disable this thread. It was causing false button presses on my USB DInput Logitech PS3 controller when emulating the classic controller on Wii. It was driving me mad so I decided to track down the exact update that introduced this bug.

**Comment by mathieui** [comment]
see pr #2744

**Pull request merged**